File: src/test/java/com/pubnub/api/endpoints/presence/HereNowEndpointTest.java
package com.pubnub.api.endpoints.presence;

import com.github.tomakehurst.wiremock.junit.WireMockRule;
import com.github.tomakehurst.wiremock.verification.LoggedRequest;
import com.pubnub.api.PubNub;
import com.pubnub.api.PubNubException;
import com.pubnub.api.callbacks.PNCallback;
import com.pubnub.api.endpoints.TestHarness;
import com.pubnub.api.enums.PNOperationType;
import com.pubnub.api.models.consumer.PNStatus;
import com.pubnub.api.models.consumer.presence.PNHereNowResult;
import org.awaitility.Awaitility;
import org.jetbrains.annotations.NotNull;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
import static com.github.tomakehurst.wiremock.client.WireMock.findAll;
import static com.github.tomakehurst.wiremock.client.WireMock.get;
import static com.github.tomakehurst.wiremock.client.WireMock.getRequestedFor;
import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
import static com.github.tomakehurst.wiremock.client.WireMock.urlMatching;
import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
import static com.github.tomakehurst.wiremock.core.WireMockConfiguration.options;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

public class HereNowEndpointTest extends TestHarness {

    @Rule
    public WireMockRule wireMockRule = new WireMockRule(options().port(this.PORT), false);

    private PubNub pubnub;
    private HereNow partialHereNow;

    @Before
    public void beforeEach() throws IOException, PubNubException {
        pubnub = this.createPubNubInstance();
        partialHereNow = pubnub.hereNow();
        wireMockRule.start();
    }

    @After
    public void afterEach() {
        pubnub.destroy();
        pubnub = null;
        wireMockRule.stop();
    }

    @Test
    public void testMultipleChannelStateSync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2"))
                .willReturn(aResponse().withBody("{\"status\":200,\"message\":\"OK\"," +
                        "\"payload\":{\"total_occupancy\":3,\"total_channels\":2," +
                        "\"channels\":{\"ch1\":{\"occupancy\":1,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}}]},\"ch2\":{\"occupancy\":2,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}},{\"uuid\":\"user3\",\"state\":{\"age\":30}}]}}}," +
                        "\"service\":\"Presence\"}")));

        PNHereNowResult response = partialHereNow.channels(Arrays.asList("ch1", "ch2")).includeState(true).sync();

        assert response != null;

        assertEquals(response.getTotalChannels(), 2);
        assertEquals(response.getTotalOccupancy(), 3);

        assertEquals(response.getChannels().get("ch1").getChannelName(), "ch1");
        assertEquals(response.getChannels().get("ch1").getOccupancy(), 1);
        assertEquals(response.getChannels().get("ch1").getOccupants().size(), 1);
        assertEquals(response.getChannels().get("ch1").getOccupants().get(0).getUuid(), "user1");
        assertEquals(response.getChannels().get("ch1").getOccupants().get(0).getState().toString(), "{\"age\":10}");

        assertEquals(response.getChannels().get("ch2").getChannelName(), "ch2");
        assertEquals(response.getChannels().get("ch2").getOccupancy(), 2);
        assertEquals(response.getChannels().get("ch2").getOccupants().size(), 2);
        assertEquals(response.getChannels().get("ch2").getOccupants().get(0).getUuid(), "user1");
        assertEquals(response.getChannels().get("ch2").getOccupants().get(0).getState().toString(), "{\"age\":10}");
        assertEquals(response.getChannels().get("ch2").getOccupants().get(1).getUuid(), "user3");
        assertEquals(response.getChannels().get("ch2").getOccupants().get(1).getState().toString(), "{\"age\":30}");

        List<LoggedRequest> requests = findAll(getRequestedFor(urlMatching("/.*")));
        assertEquals(1, requests.size());
        assertEquals("1", requests.get(0).queryParameter("state").firstValue());
    }

    @Test
    public void testMultipleChannelSync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2"))
                .willReturn(aResponse().withBody("{\"status\":200,\"message\":\"OK\"," +
                        "\"payload\":{\"total_occupancy\":3,\"total_channels\":2," +
                        "\"channels\":{\"ch1\":{\"occupancy\":1,\"uuids\":[{\"uuid\":\"user1\"}]}," +
                        "\"ch2\":{\"occupancy\":2,\"uuids\":[{\"uuid\":\"user1\"},{\"uuid\":\"user3\"}]}}}," +
                        "\"service\":\"Presence\"}")));

        PNHereNowResult response = partialHereNow.channels(Arrays.asList("ch1", "ch2")).includeState(true).sync();

        assert response != null;

        assertEquals(response.getTotalChannels(), 2);
        assertEquals(response.getTotalOccupancy(), 3);

        assertEquals(response.getChannels().get("ch1").getChannelName(), "ch1");
        assertEquals(response.getChannels().get("ch1").getOccupancy(), 1);
        assertEquals(response.getChannels().get("ch1").getOccupants().size(), 1);
        assertEquals(response.getChannels().get("ch1").getOccupants().get(0).getUuid(), "user1");
        assertNull(response.getChannels().get("ch1").getOccupants().get(0).getState());

        assertEquals(response.getChannels().get("ch2").getChannelName(), "ch2");
        assertEquals(response.getChannels().get("ch2").getOccupancy(), 2);
        assertEquals(response.getChannels().get("ch2").getOccupants().size(), 2);
        assertEquals(response.getChannels().get("ch2").getOccupants().get(0).getUuid(), "user1");
        assertNull(response.getChannels().get("ch2").getOccupants().get(0).getState());
        assertEquals(response.getChannels().get("ch2").getOccupants().get(1).getUuid(), "user3");
        assertNull(response.getChannels().get("ch2").getOccupants().get(1).getState());

        List<LoggedRequest> requests = findAll(getRequestedFor(urlMatching("/.*")));
        assertEquals(1, requests.size());
        assertEquals("1", requests.get(0).queryParameter("state").firstValue());
    }

    @Test
    public void testMultipleChannelWithoutStateSync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1,game2"))
                .willReturn(aResponse().withBody("{\"status\": 200, \"message\": \"OK\", \"payload\": {\"channels\": " +
                        "{\"game1\": {\"uuids\": [\"a3ffd012-a3b9-478c-8705-64089f24d71e\"], \"occupancy\": 1}}, " +
                        "\"total_channels\": 1, \"total_occupancy\": 1}, \"service\": \"Presence\"}")));

        PNHereNowResult response = partialHereNow.channels(Arrays.asList("game1", "game2")).includeState(false).sync();

        assert response != null;

        assertEquals(response.getTotalChannels(), 1);
        assertEquals(response.getTotalOccupancy(), 1);

        assertEquals(response.getChannels().get("game1").getChannelName(), "game1");
        assertEquals(response.getChannels().get("game1").getOccupancy(), 1);
        assertEquals(response.getChannels().get("game1").getOccupants().size(), 1);
        assertEquals(response.getChannels().get("game1").getOccupants().get(0).getUuid(), "a3ffd012-a3b9-478c-8705" +
                "-64089f24d71e");
        assertNull(response.getChannels().get("game1").getOccupants().get(0).getState());

    }

    @Test
    public void testMultipleChannelWithoutStateUUIDsSync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1,game2"))
                .willReturn(aResponse().withBody("{\"status\": 200, \"message\": \"OK\", \"payload\": {\"channels\": " +
                        "{\"game1\": {\"occupancy\": 1}}, \"total_channels\": 1, \"total_occupancy\": 1}, " +
                        "\"service\": \"Presence\"}")));

        PNHereNowResult response =
                partialHereNow.channels(Arrays.asList("game1", "game2")).includeState(false).includeUUIDs(false).sync();

        assert response != null;

        assertEquals(response.getTotalChannels(), 1);
        assertEquals(response.getTotalOccupancy(), 1);

        assertEquals(response.getChannels().get("game1").getChannelName(), "game1");
        assertEquals(response.getChannels().get("game1").getOccupancy(), 1);
        assertNull(response.getChannels().get("game1").getOccupants());

        List<LoggedRequest> requests = findAll(getRequestedFor(urlMatching("/.*")));
        assertEquals(1, requests.size());
        assertEquals("1", requests.get(0).queryParameter("disable_uuids").firstValue());
    }

    @Test
    public void testSingularChannelWithoutStateUUIDsSync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1"))
                .willReturn(aResponse().withBody("{\"status\": 200, \"message\": \"OK\", \"service\": \"Presence\", " +
                        "\"occupancy\": 3}")));

        PNHereNowResult response =
                partialHereNow.channels(Arrays.asList("game1")).includeState(false).includeUUIDs(false).sync();

        assert response != null;

        assertEquals(response.getTotalChannels(), 1);
        assertEquals(response.getTotalOccupancy(), 3);

        List<LoggedRequest> requests = findAll(getRequestedFor(urlMatching("/.*")));
        assertEquals(1, requests.size());
        assertEquals("1", requests.get(0).queryParameter("disable_uuids").firstValue());

    }

    @Test
    public void testSingularChannelWithoutStateSync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1"))
                .willReturn(aResponse().withBody("{\"status\": 200, \"message\": \"OK\", \"service\": \"Presence\", " +
                        "\"uuids\": [\"a3ffd012-a3b9-478c-8705-64089f24d71e\"], \"occupancy\": 1}")));

        PNHereNowResult response = partialHereNow.channels(Arrays.asList("game1")).includeState(false).sync();

        assert response != null;

        assertEquals(response.getTotalChannels(), 1);
        assertEquals(response.getTotalOccupancy(), 1);
        assertEquals(response.getChannels().size(), 1);
        assertEquals(response.getChannels().get("game1").getOccupancy(), 1);
        assertEquals(response.getChannels().get("game1").getOccupants().size(), 1);
        assertEquals(response.getChannels().get("game1").getOccupants().get(0).getUuid(), "a3ffd012-a3b9-478c-8705" +
                "-64089f24d71e");
        assertEquals(response.getChannels().get("game1").getOccupants().get(0).getState(), null);

    }

    @Test
    public void testSingularChannelSync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1"))
                .willReturn(aResponse().withBody("{\"status\":200,\"message\":\"OK\",\"service\":\"Presence\"," +
                        "\"uuids\":[{\"uuid\":\"a3ffd012-a3b9-478c-8705-64089f24d71e\",\"state\":{\"age\":10}}]," +
                        "\"occupancy\":1}")));

        PNHereNowResult response = partialHereNow.channels(Arrays.asList("game1")).includeState(true).sync();

        assert response != null;

        assertEquals(response.getTotalChannels(), 1);
        assertEquals(response.getTotalOccupancy(), 1);
        assertEquals(response.getChannels().size(), 1);
        assertEquals(response.getChannels().get("game1").getOccupancy(), 1);
        assertEquals(response.getChannels().get("game1").getOccupants().size(), 1);
        assertEquals(response.getChannels().get("game1").getOccupants().get(0).getUuid(), "a3ffd012-a3b9-478c-8705" +
                "-64089f24d71e");
        assertEquals(response.getChannels().get("game1").getOccupants().get(0).getState().toString(), "{\"age\":10}");

        List<LoggedRequest> requests = findAll(getRequestedFor(urlMatching("/.*")));
        assertEquals(1, requests.size());
        assertEquals("1", requests.get(0).queryParameter("state").firstValue());
    }

    @Test
    public void testSingularChannelAndGroupSync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1"))
                .willReturn(aResponse().withBody("{\"status\":200,\"message\":\"OK\",\"payload\":{\"channels\":{}, " +
                        "\"total_channels\":0, \"total_occupancy\":0},\"service\":\"Presence\"}")));

        PNHereNowResult response =
                partialHereNow.channelGroups(Arrays.asList("grp1")).channels(Arrays.asList("game1")).includeState(true).sync();

        assert response != null;

        assertEquals(response.getTotalOccupancy(), 0);
    }


    @Test
    public void testIsAuthRequiredSuccessSync() throws IOException, PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2"))
                .willReturn(aResponse().withBody("{\"status\":200,\"message\":\"OK\"," +
                        "\"payload\":{\"total_occupancy\":3,\"total_channels\":2," +
                        "\"channels\":{\"ch1\":{\"occupancy\":1,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}}]},\"ch2\":{\"occupancy\":2,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}},{\"uuid\":\"user3\",\"state\":{\"age\":30}}]}}}," +
                        "\"service\":\"Presence\"}")));

        pubnub.getConfiguration().setAuthKey("myKey");
        partialHereNow.channels(Arrays.asList("ch1", "ch2")).includeState(true).sync();

        List<LoggedRequest> requests = findAll(getRequestedFor(urlMatching("/.*")));
        assertEquals(1, requests.size());
        assertEquals("myKey", requests.get(0).queryParameter("auth").firstValue());
    }

    @Test
    public void testOperationTypeSuccessAsync() throws IOException, PubNubException, InterruptedException {
        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2"))
                .willReturn(aResponse().withBody("{\"status\":200,\"message\":\"OK\"," +
                        "\"payload\":{\"total_occupancy\":3,\"total_channels\":2," +
                        "\"channels\":{\"ch1\":{\"occupancy\":1,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}}]},\"ch2\":{\"occupancy\":2,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}},{\"uuid\":\"user3\",\"state\":{\"age\":30}}]}}}," +
                        "\"service\":\"Presence\"}")));

        final AtomicInteger atomic = new AtomicInteger(0);

        partialHereNow.async(new PNCallback<PNHereNowResult>() {
            @Override
            public void onResponse(PNHereNowResult result, @NotNull PNStatus status) {
                if (status != null && status.getOperation() == PNOperationType.PNHereNowOperation) {
                    atomic.incrementAndGet();
                }

            }
        });

        Awaitility.await().atMost(5, TimeUnit.SECONDS).untilAtomic(atomic, org.hamcrest.core.IsEqual.equalTo(1));
    }

    @Test(expected = PubNubException.class)
    public void testNullSubKeySync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2"))
                .willReturn(aResponse().withBody("{\"status\":200,\"message\":\"OK\"," +
                        "\"payload\":{\"total_occupancy\":3,\"total_channels\":2," +
                        "\"channels\":{\"ch1\":{\"occupancy\":1,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}}]},\"ch2\":{\"occupancy\":2,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}},{\"uuid\":\"user3\",\"state\":{\"age\":30}}]}}}," +
                        "\"service\":\"Presence\"}")));

        pubnub.getConfiguration().setSubscribeKey(null);
        partialHereNow.channels(Arrays.asList("ch1", "ch2")).includeState(true).sync();
    }

    @Test(expected = PubNubException.class)
    public void testEmptySubKeySync() throws PubNubException, InterruptedException {

        stubFor(get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2"))
                .willReturn(aResponse().withBody("{\"status\":200,\"message\":\"OK\"," +
                        "\"payload\":{\"total_occupancy\":3,\"total_channels\":2," +
                        "\"channels\":{\"ch1\":{\"occupancy\":1,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}}]},\"ch2\":{\"occupancy\":2,\"uuids\":[{\"uuid\":\"user1\"," +
                        "\"state\":{\"age\":10}},{\"uuid\":\"user3\",\"state\":{\"age\":30}}]}}}," +
                        "\"service\":\"Presence\"}")));

        pubnub.getConfiguration().setSubscribeKey("");
        partialHereNow.channels(Arrays.asList("ch1", "ch2")).includeState(true).sync();
    }

}

File: src/test/kotlin/com/pubnub/api/endpoints/presence/HereNowEndpointTest.kt
package com.pubnub.api.endpoints.presence

import com.github.tomakehurst.wiremock.client.WireMock.*
import com.pubnub.api.BaseTest
import com.pubnub.api.PubNubError.SUBSCRIBE_KEY_MISSING
import com.pubnub.api.PubNubException
import com.pubnub.api.enums.PNOperationType.PNHereNowOperation
import com.pubnub.api.models.consumer.presence.PNHereNowResult
import org.awaitility.Awaitility
import org.hamcrest.core.IsEqual
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeUnit.SECONDS
import java.util.concurrent.atomic.AtomicInteger

class HereNowEndpointTest : BaseTest() {

    @Test
    fun testMultipleChannelStateSync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2")).willReturn(
                aResponse().withBody(
                    """
                    {
                      "status": 200,
                      "message": "OK",
                      "payload": {
                        "total_occupancy": 3,
                        "total_channels": 2,
                        "channels": {
                          "ch1": {
                            "occupancy": 1,
                            "uuids": [
                              {
                                "uuid": "user1",
                                "state": {
                                  "age": 10
                                }
                              }
                            ]
                          },
                          "ch2": {
                            "occupancy": 2,
                            "uuids": [
                              {
                                "uuid": "user1",
                                "state": {
                                  "age": 10
                                }
                              },
                              {
                                "uuid": "user3",
                                "state": {
                                  "age": 30
                                }
                              }
                            ]
                          }
                        }
                      },
                      "service": "Presence"
                    }
                """.trimIndent()
                )
            )
        )

        val response = pubnub.hereNow().apply {
            channels = listOf("ch1", "ch2")
            includeState = true
        }.sync()!!

        assertEquals(response.totalChannels, 2)
        assertEquals(response.totalOccupancy, 3)

        assertEquals(response.channels["ch1"]!!.channelName, "ch1")
        assertEquals(response.channels["ch1"]!!.occupancy, 1)
        assertEquals(response.channels["ch1"]!!.occupants.size, 1)
        assertEquals(response.channels["ch1"]!!.occupants[0].uuid, "user1")
        assertEquals(response.channels["ch1"]!!.occupants[0].state.toString(), """{"age":10}""")

        assertEquals(response.channels["ch2"]!!.channelName, "ch2")
        assertEquals(response.channels["ch2"]!!.occupancy, 2)
        assertEquals(response.channels["ch2"]!!.occupants.size, 2)
        assertEquals(response.channels["ch2"]!!.occupants[0].uuid, "user1")
        assertEquals(response.channels["ch2"]!!.occupants[0].state.toString(), """{"age":10}""")
        assertEquals(response.channels["ch2"]!!.occupants[1].uuid, "user3")
        assertEquals(response.channels["ch2"]!!.occupants[1].state.toString(), """{"age":30}""")

        val requests = findAll(getRequestedFor(urlMatching("/.*")))
        assertEquals(1, requests.size)
        assertEquals("1", requests[0].queryParameter("state").firstValue())
    }

    @Test
    fun testMultipleChannelSync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2"))
                .willReturn(
                    aResponse().withBody(
                        """
                            {
                              "status": 200,
                              "message": "OK",
                              "payload": {
                                "total_occupancy": 3,
                                "total_channels": 2,
                                "channels": {
                                  "ch1": {
                                    "occupancy": 1,
                                    "uuids": [
                                      {
                                        "uuid": "user1"
                                      }
                                    ]
                                  },
                                  "ch2": {
                                    "occupancy": 2,
                                    "uuids": [
                                      {
                                        "uuid": "user1"
                                      },
                                      {
                                        "uuid": "user3"
                                      }
                                    ]
                                  }
                                }
                              },
                              "service": "Presence"
                            }
                        """.trimIndent()
                    )
                )
        )

        val response = pubnub.hereNow().apply {
            channels = listOf("ch1", "ch2")
            includeState = true
        }.sync()!!

        assertEquals(response.totalChannels, 2)
        assertEquals(response.totalOccupancy, 3)

        assertEquals(response.channels["ch1"]!!.channelName, "ch1")
        assertEquals(response.channels["ch1"]!!.occupancy, 1)
        assertEquals(response.channels["ch1"]!!.occupants.size, 1)
        assertEquals(response.channels["ch1"]!!.occupants[0].uuid, "user1")
        assertNull(response.channels["ch1"]!!.occupants[0].state)

        assertEquals(response.channels["ch2"]!!.channelName, "ch2")
        assertEquals(response.channels["ch2"]!!.occupancy, 2)
        assertEquals(response.channels["ch2"]!!.occupants.size, 2)
        assertEquals(response.channels["ch2"]!!.occupants[0].uuid, "user1")
        assertNull(response.channels["ch2"]!!.occupants[0].state)
        assertEquals(response.channels["ch2"]!!.occupants[1].uuid, "user3")
        assertNull(response.channels["ch2"]!!.occupants[1].state)

        val requests = findAll(getRequestedFor(anyUrl()))
        assertEquals(1, requests.size)
        assertEquals("1", requests[0].queryParameter("state").firstValue())
    }

    @Test
    fun testMultipleChannelWithoutStateSync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1,game2"))
                .willReturn(
                    aResponse().withBody(
                        """
                            {
                              "status": 200,
                              "message": "OK",
                              "payload": {
                                "channels": {
                                  "game1": {
                                    "uuids": [
                                      "a3ffd012-a3b9-478c-8705-64089f24d71e"
                                    ],
                                    "occupancy": 1
                                  }
                                },
                                "total_channels": 1,
                                "total_occupancy": 1
                              },
                              "service": "Presence"
                            }
                        """.trimIndent()
                    )
                )
        )

        val response = pubnub.hereNow().apply {
            channels = listOf("game1", "game2")
            includeState = false
        }.sync()!!

        assertEquals(response.totalChannels, 1)
        assertEquals(response.totalOccupancy, 1)

        assertEquals(response.channels["game1"]!!.channelName, "game1")
        assertEquals(response.channels["game1"]!!.occupancy, 1)
        assertEquals(response.channels["game1"]!!.occupants.size, 1)
        assertEquals(response.channels["game1"]!!.occupants[0].uuid, "a3ffd012-a3b9-478c-8705-64089f24d71e")
        assertNull(response.channels["game1"]!!.occupants[0].state)
    }

    @Test
    fun testMultipleChannelWithoutStateUUIDsSync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1,game2"))
                .willReturn(
                    aResponse().withBody(
                        """
                        {
                          "status": 200,
                          "message": "OK",
                          "payload": {
                            "channels": {
                              "game1": {
                                "occupancy": 1
                              }
                            },
                            "total_channels": 1,
                            "total_occupancy": 1
                          },
                          "service": "Presence"
                        }
                    """.trimIndent()
                    )
                )
        )

        val response = pubnub.hereNow().apply {
            channels = listOf("game1", "game2")
            includeState = false
            includeUUIDs = false
        }.sync()!!

        assertEquals(response.totalChannels, 1)
        assertEquals(response.totalOccupancy, 1)

        assertEquals(response.channels["game1"]!!.channelName, "game1")
        assertEquals(response.channels["game1"]!!.occupancy, 1)
        assertTrue(response.channels["game1"]!!.occupants.isEmpty())

        val requests = findAll(getRequestedFor(urlMatching("/.*")))
        assertEquals(1, requests.size)
        assertEquals("1", requests[0].queryParameter("disable_uuids").firstValue())
    }

    @Test
    fun testSingularChannelWithoutStateUUIDsSync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1"))
                .willReturn(
                    aResponse().withBody(
                        """
                            {
                              "status": 200,
                              "message": "OK",
                              "service": "Presence",
                              "occupancy": 3
                            }
                        """.trimIndent()
                    )
                )
        )

        val response = pubnub.hereNow().apply {
            channels = listOf("game1")
            includeUUIDs = false
            includeUUIDs = false
        }.sync()!!

        println(response)

        assertEquals(response.totalChannels, 1)
        assertEquals(response.totalOccupancy, 3)
        println("response.channels: ${response.channels}")

        val requests = findAll(getRequestedFor(urlMatching("/.*")))
        assertEquals(1, requests.size)
        assertEquals("1", requests[0].queryParameter("disable_uuids").firstValue())
    }

    @Test
    fun testSingularChannelWithoutStateSync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1"))
                .willReturn(
                    aResponse().withBody(
                        """
                            {
                              "status": 200,
                              "message": "OK",
                              "service": "Presence",
                              "uuids": [
                                "a3ffd012-a3b9-478c-8705-64089f24d71e"
                              ],
                              "occupancy": 1
                            }
                        """.trimIndent()
                    )
                )
        )

        val response = pubnub.hereNow().apply {
            channels = listOf("game1")
            includeState = false
        }.sync()!!

        assertEquals(response.totalChannels, 1)
        assertEquals(response.totalOccupancy, 1)
        assertEquals(response.channels.size, 1)
        assertEquals(response.channels["game1"]!!.occupancy, 1)
        assertEquals(response.channels["game1"]!!.occupants.size, 1)
        assertEquals(response.channels["game1"]!!.occupants[0].uuid, "a3ffd012-a3b9-478c-8705-64089f24d71e")
        assertEquals(response.channels["game1"]!!.occupants[0].state, null)
    }

    @Test
    fun testSingularChannelSync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1"))
                .willReturn(
                    aResponse().withBody(
                        """
                            {
                              "status": 200,
                              "message": "OK",
                              "service": "Presence",
                              "uuids": [
                                {
                                  "uuid": "a3ffd012-a3b9-478c-8705-64089f24d71e",
                                  "state": {
                                    "age": 10
                                  }
                                }
                              ],
                              "occupancy": 1
                            }
                        """.trimIndent()
                    )
                )
        )
        val response = pubnub.hereNow().apply {
            channels = listOf("game1")
            includeState = true
        }.sync()!!

        assertEquals(response.totalChannels, 1)
        assertEquals(response.totalOccupancy, 1)
        assertEquals(response.channels.size, 1)
        assertEquals(response.channels["game1"]!!.occupancy, 1)
        assertEquals(response.channels["game1"]!!.occupants.size, 1)
        assertEquals(response.channels["game1"]!!.occupants[0].uuid, "a3ffd012-a3b9-478c-8705-64089f24d71e")
        assertEquals(response.channels["game1"]!!.occupants[0].state.toString(), """{"age":10}""")
        val requests = findAll(getRequestedFor(urlMatching("/.*")))
        assertEquals(1, requests.size)
        assertEquals("1", requests[0].queryParameter("state").firstValue())
    }

    @Test
    fun testSingularChannelAndGroupSync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/game1"))
                .willReturn(
                    aResponse().withBody(
                        """
                            {
                              "status": 200,
                              "message": "OK",
                              "payload": {
                                "channels": {},
                                "total_channels": 0,
                                "total_occupancy": 0
                              },
                              "service": "Presence"
                            }
                        """.trimIndent()
                    )
                )
        )

        val response = pubnub.hereNow().apply {
            channels = listOf("game1")
            channelGroups = listOf("grp1")
            includeState = true
        }.sync()!!

        assertEquals(response.totalOccupancy, 0)
    }

    @Test
    fun testIsAuthRequiredSuccessSync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2"))
                .willReturn(
                    aResponse().withBody(
                        """
                            {
                              "status": 200,
                              "message": "OK",
                              "payload": {
                                "total_occupancy": 3,
                                "total_channels": 2,
                                "channels": {
                                  "ch1": {
                                    "occupancy": 1,
                                    "uuids": [
                                      {
                                        "uuid": "user1",
                                        "state": {
                                          "age": 10
                                        }
                                      }
                                    ]
                                  },
                                  "ch2": {
                                    "occupancy": 2,
                                    "uuids": [
                                      {
                                        "uuid": "user1",
                                        "state": {
                                          "age": 10
                                        }
                                      },
                                      {
                                        "uuid": "user3",
                                        "state": {
                                          "age": 30
                                        }
                                      }
                                    ]
                                  }
                                }
                              },
                              "service": "Presence"
                            }
                        """.trimIndent()
                    )
                )
        )

        pubnub.configuration.authKey = "myKey"

        pubnub.hereNow().apply {
            channels = listOf("ch1", "ch2")
            includeState = true
        }.sync()!!

        val requests = findAll(getRequestedFor(urlMatching("/.*")))
        assertEquals(1, requests.size)
        assertEquals("myKey", requests[0].queryParameter("auth").firstValue())
    }

    @Test
    fun testOperationTypeSuccessAsync() {
        stubFor(
            get(urlPathEqualTo("/v2/presence/sub_key/mySubscribeKey/channel/ch1,ch2"))
                .willReturn(
                    aResponse().withBody(
                        """
                            {
                              "status": 200,
                              "message": "OK",
                              "payload": {
                                "total_occupancy": 3,
                                "total_channels": 2,
                                "channels": {
                                  "ch1": {
                                    "occupancy": 1,
                                    "uuids": [
                                      {
                                        "uuid": "user1",
                                        "state": {
                                          "age": 10
                                        }
                                      }
                                    ]
                                  },
                                  "ch2": {
                                    "occupancy": 2,
                                    "uuids": [
                                      {
                                        "uuid": "user1",
                                        "state": {
                                          "age": 10
                                        }
                                      },
                                      {
                                        "uuid": "user3",
                                        "state": {
                                          "age": 30
                                        }
                                      }
                                    ]
                                  }
                                }
                              },
                              "service": "Presence"
                            }
                        """.trimIndent()
                    )
                )
        )

        val atomic = AtomicInteger(0)

        pubnub.hereNow().async { _, status ->
            if (status.operation == PNHereNowOperation) {
                atomic.incrementAndGet();
            }
        }

        Awaitility.await().atMost(5, SECONDS).untilAtomic(atomic, IsEqual.equalTo(1))
    }

    @Test
    fun testEmptySubKeySync() {
        pubnub.configuration.subscribeKey = ""

        var response: PNHereNowResult? = null
        try {
            response = pubnub.hereNow().apply {
                channels = listOf("ch1", "ch2")
                includeState = true
            }.sync()
        } catch (e: Exception) {
            assertNull(response)
            assertTrue((e as PubNubException).pubnubError == SUBSCRIBE_KEY_MISSING)
        }
    }

}